jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -850,6 +856,11 @@ public String(char[] data, int start, int length) {
 				}
 
 				coder = LATIN1;
+			} else if (COMPACT_STRINGS && ((start != 0) || (data.length != length * 2)) && helpers.canEncodeAsLatin1(data, start, length)) {
```

Comment:
I don't think we want `((start != 0) || (data.length != length * 2))` included in this condition: we want to use the compressed representation whether that's true or not.

Reply:
The idea is that if the original string isn't compressed and the size doesn't change, then the new string can't be compressed either so checking it is a waste of time. Atm I believe all the callers abide by this rule.

Reply:
There is also an optimization in the next else case where the byte[] is shared rather than creating a copy, which we don't want to loose.

Reply:
With one exception, all uses of the affected constructors (where `length != 1`) supply `start != 0` or `data.length != length * 2`, so testing that condition again here is the waste of time.

That one exception is `replaceAll()` where the `regex` is one (perhaps escaped) character and the replacement is at most one character. If the original string was compressed, the code above (lines 849-858) reuses the array as desired. If the original string was not compressed, the new code that calls `StringUTF16.newString()` handles creating a compressed string when appropriate.

Reply:
Actually I'd introduced a bug in replaceAll(), it can't call String(byte[], UTF16) because this doesn't take the new length into account. Updated as suggested.

Reply:
Added back the String(byte[], UTF16) opt, but corrected the condition when it's used.

Reply:
Removed the opts in replace/replaceAll again, since the code doesn't track if the replacement character is actually used.